buildSrc/subprojects/parser/src/main/java/org/apache/calcite/buildtools/parser/DialectTraverser.java

```diff
@@ -141,21 +156,47 @@ private void traverse(
     String nextDirectory = directories.peek();
     for (File f : files) {
       String fileName = f.getName();
-      if (f.isFile() && fileName.endsWith(".ftl")) {
+      if (f.isFile()) {
```

Comment:
Can we reverse the polarity of this if-check and continue early? That would allow us to unnest all of this logic, which appears to be larger than that in the else-if.

Reply:
I'm not sure if that is possible given that there is an else-if after it.

Reply:
Why can't we do the following?

`
if (!directories.isEmpty() && fileName.equals(nextDirectory)) {
  ...
  continue;
}
<code that was previously in if-clause>
`

Reply:
This seems to be breaking when the traverser encounters a directory that isn't the `nextDirectory`.

Reply:
I see. Then you should be able to fix that by updating the if-clause to:

`if (!f.isFile() && !directories.isEmpty() && fileName.equals(nextDirectory)) { ... continue; }`

This is just an optional suggestion. Just a thought that may make it a bit easier to read this by reducing the amount of nesting.

Reply:
I also tried that and it fails for the same reason since that if statement can be false both when `f` is a directory and `f` is a file.

Reply:
Hm, not sure I fully follow. But it doesn't really matter either way. Feel free to leave it as is for now.